### PR TITLE
test_: added tests for updating fees when after router provides the best route

### PR DESCRIPTION
**Pull request by saledjenic**
This PR introduces anvil network on the status-go side and tests:
- subscribing/unsubscribing for fees updates
- receiving fees updates

### CODE REVIEW

File: api/backend_test.go [hunk 1]
```diff
@@ -90,50 +90,41 @@ func setupTestMultiDB() (*multiaccounts.Database, func() error, error) {
...
+		return nil, nil, err
 	}
 	backend.StatusNode().SetWalletDB(walletDb)
 
-	return backend, stop1, stop2, stop3, err
+	return backend, func(t *testing.T) error {
+		require.NoError(t, stop1())
+		require.NoError(t, stop2())
+		require.NoError(t, stop3())
+		require.NoError(t, backend.StopNode())
```
Comment (igor-sirotin) [inline api/backend_test.go:117]: You can use `t.Cleanup` for this:
```go
t.Cleanup(func() {
    stop1()
    stop2()
    stop3()
    backend.StopNode()
})
```

And no need to return custom callback then 🙌

File: api/wallet_router_integration_test.go [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+			assert.NoError(t, err)
+
+			suggestedRoutesCh <- response.Routes
+		}
+	})
+	signal.SetMobileSignalHandler(signalHandler)
+
+	closeFn := func() {
+		close(suggestedRoutesCh)
+		signal.SetMobileSignalHandler(nil)
```
Comment (igor-sirotin) [inline api/wallet_router_integration_test.go]: Thanks for keeping an eye on the reset of `MobileSignalHandler` ❤️ 

I recently added a `ResetMobileSignalHandler`, please checkout [this](https://www.notion.so/Code-Smell-6-Cleanup-MobileSignalHandler-67dda79e7e0f4275b105af557d890b8d?pvs=4)

TLDR: Just add `t.Cleanup(signal.ResetMobileSignalHandler)` instead, it's a bit simpler.

File: api/wallet_router_integration_test.go [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+	return suggestedRoutesCh, closeFn
+}
+
+func setupWalletWithAnvil(t *testing.T, mnemonic string, password string, rootDataDir string) (*GethStatusBackend, func(t *testing.T) error, error) {
+	const pathEIP1581Root = "m/43'/60'/1581'"
+	const pathEIP1581Chat = pathEIP1581Root + "/0'/0"
+	const pathWalletRoot = "m/44'/60'/0'/0"
+	const pathDefaultWalletAccount = pathWalletRoot + "/0"
+	var allGeneratedPaths []string
+	allGeneratedPaths = append(allGeneratedPaths, pathEIP1581Root, pathEIP1581Chat, pathWalletRoot, pathDefaultWalletAccount)
```
Comment (igor-sirotin) [inline api/wallet_router_integration_test.go]: 🤔 
```suggestion
	allGeneratedPaths := []string{pathEIP1581Root, pathEIP1581Chat, pathWalletRoot, pathDefaultWalletAccount}
```

File: api/wallet_router_integration_test.go [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+		KeyUID:    profileKeypair.KeyUID,
+	}
+
+	err = backend.ensureAppDBOpened(account, password)
+	require.NoError(t, err)
+
+	err = backend.StartNodeWithAccountAndInitialConfig(account, password, *defaultSettings, nodeConfig, profileKeypair.Accounts, nil)
+	require.NoError(t, err)
+
+	return backend, stop, nil
```
Comment (igor-sirotin) [inline api/wallet_router_integration_test.go]: Same here for `stop`, using `t.Cleanup` help

File: api/wallet_router_integration_test.go (hunk 2 above)
Comment (saledjenic) [reply api/wallet_router_integration_test.go]: Updated.

File: api/wallet_router_integration_test.go (hunk 3 above)
Comment (saledjenic) [reply api/wallet_router_integration_test.go]: Updated.

File: api/backend_test.go (hunk 1 above)
Comment (saledjenic) [reply api/backend_test.go:117]: Updated.

File: api/wallet_router_integration_test.go (hunk 4 above)
Comment (saledjenic) [reply api/wallet_router_integration_test.go]: Updated.

**Comment by saledjenic** [comment]
@igor-sirotin thanks for the review, and comments addressed.

**Comment by igor-sirotin** [comment]
Tests seem to be broken, investigating

**Review by igor-sirotin — CHANGES_REQUESTED** [review]
Don't take this personally, just to ensure it's not merged while the tests are failing 😄 
Details: https://github.com/status-im/status-go/pull/5797

**Comment by saledjenic** [comment]
> Don't take this personally, just to ensure it's not merged while the tests are failing 😄 Details: #5797

Yes, I know that. Discussed with Anton, already. :)

**Comment by igor-sirotin** [comment]
Accidents happen, you know 😄

**Comment by saledjenic** [comment]
@igor-sirotin @antdanchenko please take a look at the tests now. 
I moved updates `TestRouterFeesUpdate` to another package `anviltests`. We can use that package for all integration tests that need anvil running, and also I excluded that package from running unit tests. Also added `run-integration-anvil` to makefile, and what we need now is just to add that call to CI.

**Review by alaibe — APPROVED** [review]

**Review by jakubgs — APPROVED** [review]
`_assets/ci/Jenkinsfile.tests-anvil` is awfully similar to `_assets/ci/Jenkinsfile.tests-rpc`. But I'm assuming it's going to change in the future.

File: integration-tests/docker-compose.anvil.yml [hunk 5]
```diff
@@ -3,7 +3,7 @@ services:
     image: ghcr.io/foundry-rs/foundry:latest
     platform: linux/amd64
     command:
-      - anvil --host 0.0.0.0
+      - anvil --host 0.0.0.0 --block-time 2
```
Comment (antdanchenko) [inline integration-tests/docker-compose.anvil.yml]: we don't need this change for integration-tests

File: integration-tests-anvil/docker-compose.yml [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+
+  go-test:
+    image: golang:1.21.4
+    depends_on:
+      - anvil
+    volumes:
+      - ../:/go/src/github.com/status-im/status-go/
+    working_dir: /go/src/github.com/status-im/status-go/api
+    command: |
+        go test -tags gowaku_skip_migrations -run ^TestRouterFeesUpdate github.com/status-im/status-go/api/... -v -- -anvil-host=http://anvil:8545
```
Comment (antdanchenko) [inline integration-tests-anvil/docker-compose.yml]: right now we are launching a single test here, would be great to replace it with launching all go tests that require anvil if possible

File: Makefile [hunk 7]
```diff
@@ -499,3 +499,6 @@ run-integration-tests:
 run-anvil: SHELL := /bin/sh
 run-anvil:
 	docker-compose -f integration-tests/docker-compose.anvil.yml up --remove-orphans
+
+run-integration-anvil:
```
Comment (antdanchenko) [inline Makefile:503]: CI run won't finish because anvil will continue to run forever, we need to do something similar to https://github.com/status-im/status-go/blob/develop/Makefile#L492 with running silently and then reading exit code from the container with tests

File: integration-tests-anvil/docker-compose.yml (hunk 6 above)
Comment (saledjenic) [reply integration-tests-anvil/docker-compose.yml]: Updated.

File: Makefile (hunk 7 above)
Comment (saledjenic) [reply Makefile:503]: Updated.

**Comment by igor-sirotin** [comment]
I probably should have looked at this closer in the very beginning, but better to speak late than never 🙂 
I think we're a bit messing up things here.

1. I don't think that making `defaultSettings` and `ensureDBOpened` public is a good idea.
    `ensureDBOpened` is for sure a low-level API, it should be kept private. Should not be used in integrations tests.
    
    The fact that we have to make them **public** for this is a sign of that we're doing something wrong.

2. The test itself if ~50 lines of code at the end of `TestRouterFeesUpdate`.
    Yet we have 300 more lines just to set it up.

    I believe this is also a sign of that we're doing something wrong.

3. Can't we implement `TestRouterFeesUpdate` with the `integration-tests` that we already have?
    It even imports the same seed phrase 😄 
    Looks like it would be much simpler to implement it that way.

cc @antdanchenko @saledjenic @status-im/status-go-guild 

----

These are some major questions I have. 
I still have minor concerns for the current implementation itself, I will write them if we decided to keep everything as is.

**Comment by saledjenic** [comment]
> I don't think that making defaultSettings and ensureDBOpened public is a good idea.
ensureDBOpened is for sure a low-level API, it should be kept private. Should not be used in integrations tests.

Exactly, that's what I said to you in private chat, we have to sort that out a bit better cause `defaultSettings`, `defaultNodeConfig` and other things are used in the same/similar form in many places in statusgo and I could copy them, but I didn't want to have one more place with those functions, and made them public, with the hope to have better organization and reusable function there, soon.

> The test itself if ~50 lines of code at the end of TestRouterFeesUpdate.
Yet we have 300 more lines just to set it up.
> I believe this is also a sign of that we're doing something wrong.

Yes, looks like that, but we need `rpc.Client` set with Anvil for those tests, and in this particular case `subscribeForUdates` and `unsubscribeFeesUpdateAccrossAllChains` are private to `router` package, not exposed and can be tests only via `SuggestedRoutesAsync` and `SuggestedRoutes` calls if the route is generated. In order to generate the route we need all other things to be set, like settings, node config... that's where other 300 lines came from.

> Can't we implement TestRouterFeesUpdate with the integration-tests that we already have?
It even imports the same seed phrase 😄
Looks like it would be much simpler to implement it that way.

I am not sure that we can, cause those `integration-tests` are written in python and they can call only exposed endpoints (we cannot test just updates, but the router, and that's maybe ok), but I am not sure if it's possible to connect to signal that is sent when fees are recacluated? Guess not.

The point of having them this way is that we will add more tests later here that will test the router, sending transaction process and that way we can get rid of `if params.TestsMode` checks that we temporarily introduced. Maybe for this one we can use `integration-tests` (not sure @antdanchenko ?) but we need to be sure that `rpc.Client` will be set correctly, also a full control over base fees, gas estimation, balances per different chains and so on in order to compare a real with the expected result.

----

I agree that we should think about this, so no rush in making a decision. This PR can stay like this, at least as a PoC, till we figure out what's the best approach to proceed with.

@igor-sirotin thanks for the review, there is a long way in front of us till we set everything properly.

File: _assets/ci/Jenkinsfile.tests-anvil [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+      steps { script {
+        nix.shell('make lint', pure: true)
+      } }
+    }
+
+    stage('Canary') {
+      steps { script {
+        nix.shell('make canary-test', pure: true)
+      } }
+    }
```
Comment (jakubgs) [inline _assets/ci/Jenkinsfile.tests-anvil]: Why did you guys add the same steps as in `Jenkinsfile.tests`? What's the point of that?

**Comment by igor-sirotin** [comment]
We agreed to postpone this PR. 

The main reason why we can't implement this with existing integration tests, is the disability of getting events from status-go. I opened an issue about it here: 
- https://github.com/status-im/status-go/issues/5803.

It shouldn't be difficult to implement, I will take a look at this ASAP.

cc @saledjenic @antdanchenko

**Comment by igor-sirotin** [comment]
Here's the PR with required functionality implemented:
- https://github.com/status-im/status-go/pull/5806

**Comment by saledjenic** [comment]
Closing this cause tests using Anvil chain are/will be done as functional tests, written in Python and running statusd (status-go deamon).
